Trigger symfloat specialization in argument binding code

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139933
* #139583
* #138922
* #138918
* #138666
* #138915
* #139587
* #139935
* #139896
* __->__ #139454
* #139846
* #139572
* #139568
* #139457
* #139569

Fixes the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
specialize symfloats in various places to support specialize_float=False rollout

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -2385,6 +2385,9 @@ def handle_traced_output(example_value, tx, proxy, options, subclass_type, targe
         return ConstantVariable.create(example_value, **options)
+    elif isinstance(example_value, int) and (proxy.node.target in ["__round__"]):
+        set_example_value(proxy.node, example_value)
+        return ConstantVariable.create(example_value, **options)
```
Merge this with below? If you can unconditionally let this through that would be good too lol

ezyang commented on torch/nn/functional.py:
```diff
@@ -4638,6 +4638,12 @@ def interpolate(  # noqa: F811
         )
 
+    if not torch.jit.is_scripting() and scale_factors is not None:
+        # We need to specialize symfloats for now. Eventually we should do a tensorify pass in dynamo.
```
This is a suspicious spot to do it, would have expected later

bobrenjc93 replied:
Did you have a place in mind? Generally for all of these fixes I'm specializing right before the point it needs to in the stacktrace (via debugger + jumping up and down stacks). If there's a better method to find the precise place to put these specializations, do let me know!

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -2185,6 +2185,15 @@ def get_fake_value(node, tx, allow_non_graph_fake=False):
+            if isinstance(arg, torch.SymFloat) and arg.node.hint is not None
+            else arg
+            for arg in args
+        )
```
Also a suspicious spot to do it

bobrenjc93 replied:
Did you have a place in mind? Generally for all of these fixes I'm specializing right before the point it needs to in the stacktrace (via debugger + jumping up and down stacks). If there's a better method to find the precise place to put these specializations, do let me know!

ezyang replied:
What I am generically expecting is if I pass a SymFloat to a PyTorch API that only accepts 'float', I expect that API to automatically trigger specialization in its argument binding code (as it attempts to get a float and gets a SymFloat instead.) The analogous code for SymInt was added in https://github.com/pytorch/pytorch/pull/95479 so you probably need the float version of it too.

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "specialize symfloats in various places to support specialize_float=False rollout"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Trigger symfloat specialization in argument binding code"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Trigger symfloat specialization in argument binding code"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Trigger symfloat specialization in argument binding code"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Trigger symfloat specialization in argument binding code"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

while ensuring

`python test/dynamo/test_dynamic_shapes.py DynamicShapesFunctionTests.test_number_method_method_is_integer_num_type6_dynamic_shapes` continues to pass



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Trigger symfloat specialization in argument binding code"


One day we will handle these cases via dynamo tensorify passes but for now we need to specialize. These changes in particular fix the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`

while ensuring

`python test/dynamo/test_dynamic_shapes.py DynamicShapesFunctionTests.test_number_method_method_is_integer_num_type6_dynamic_shapes` continues to pass



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -27,12 +27,14 @@ static std::unordered_map<std::string, ParameterType> type_map = {
     {"c10::List<::std::optional<Tensor>>", ParameterType::TENSOR_LIST},
     {"IntArrayRef", ParameterType::INT_LIST},
     {"SymIntArrayRef", ParameterType::SYM_INT_LIST},
+    {"SymFloatArrayRef", ParameterType::SYM_FLOAT_LIST},
```
All of this code for SymFloat and SymFloatArrayRef isn't actually doing anything, because you are never code generating SymFloat/SymFloatArrayRef entries in the string passed to Python argument parser.  You can drop all of it, you only need to change the semantics of `is_float_or_complex_list` to also accept SymFloat, and ditto with THPUtils_checkDouble to switch to is_float_or_symfloat.

bobrenjc93 replied:
Should we do the same thing for SymInt? If not, what's the difference?

ezyang reviewed (approved):
I'm approving because the bogus changes are harmless, but it is better to excise them before landing.

bobrenjc93 committed:
Update on "Trigger symfloat specialization in argument binding code"


Fixes the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Trigger symfloat specialization in argument binding code"


Fixes the test `python test/inductor/test_torchinductor.py CpuTests.test_upsample_cat_conv_cpu` when `specialize_float=False`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented:
@pytorchbot merge
